### ISSUE TITLE
SSH parameters and ProxyCommand

### ISSUE THREAD

**Issue by gregkalapos** [microsoft/azure-pipelines-tasks#4186]
Hi, 
i have an asp.net core application, which i want to build and publish with vsts to a Linux machine (currently I deploy it manually). In order to publish it the only option i have is SSH (I don't control the server, so I cannot open up other ports, etc.). The setup for the production server is this: There is a proxy, which is available via the internet and the production linux machine I deploy to is behind the proxy. So the deployment target is not available directly. In order to SSH into the production machine I use this command: 

`ssh -o ProxyCommand="ssh -i ~/.ssh/[SSHKey] -W %h:%p proxyUser@ProxyMachine" ProdMachineUser@ProdMachine`

And the way I copy files to the machine is this: 

`scp -o ProxyCommand="ssh -i ~/.ssh/[SSHKey] -W %h:%p proxyUser@ProxyMachine" ~/MyApplication.zip ProdMachineUser@ProdMachine:~/`

I was happy to see that there is already an 'SSH' and a 'Copy Files Over SSH' task. The problem is that under 'Endpoints' there is no option to enter SSH parameters (it only accepts a host, port, username, passphrase, privatekey). Unfortunately that does not seem to work with my setup, since I cannot directly log in to 'ProxyMachine', i have to jump from there without logging in. 

[Here ](https://monkeyswithbuttons.wordpress.com/2010/10/01/ssh-proxycommand-and-putty/)is a tutorial about this for putty. Something similar would be nice in vsts.

So the questions:
- Is it possible to use the existing SSH tasks with this setup? (especially with -o ProxyCommand)
- If not, is it planned to support this later? If yes, could you please point me to a documentation on this?
- What about other SSH parameters? (see http://linuxcommand.org/man_pages/ssh1.html) there is a lot more than host and port...
- Is there an ssh/config file for the tasks? that would also solve the problem, since I could set the proxy there.

Thanks!

**Comment by madhurig** [comment]
@gregkalapos: The current SSH tasks don't shell out to the command line (since it requires interactive input at least one time), the task uses the [SSH2 node module](https://github.com/mscdex/ssh2). 

- The proxy support will require connection hopping that is not currently implemented in the task. So existing task cannot be used with your setup. We will consider this an enhancement for a future version. 
- Since we don't shell out to command line, any additional parameter supports needs to be implemented in the task. We are considering those based on customer requests. Are there particular ones you would like to use?

For your immediate solution, I would recommend writing a script and using the [Shell Script task](https://www.visualstudio.com/en-us/docs/build/steps/utility/shell-script) or [Batch Script task](https://www.visualstudio.com/en-us/docs/build/steps/utility/batch-script) to do the deploy. You might need to setup the SSH keys manually one time on your build server but with a script you will have more flexibility i.e. you can pass all arguments supported by the client and also use an SSH client of your choice.

Thanks,
Madhuri

**Comment by gregkalapos** [comment]
@madhurig thanks for the response.

> Since we don't shell out to command line, any additional parameter supports needs to be implemented in the task. We are considering those based on customer requests. Are there particular ones you would like to use?

The only one, which I need currently is the '-o ProxyCommand'. 

>   I would recommend writing a script and using the Shell Script task or Batch Script task to do the deploy. You might need to setup the SSH keys manually one time on your build server...

Well, I already tried to do that, but was unable to figure out how to use my private key (which I need to get into the SSH proxy machine). So the first SSH connection (the one between the local machine and the Proxy) is done via SSH key. I use VSTS and I did not find any option to upload my private key (except checking that in, but checking in a private key into a repo is imo not the way to go). The second SSH connection (the one from the proxy to the real production machine) needs a password. So there I don't need the private key, but I have to type the passsword. 

So is there a way to store private SSH keys in VSTS (for the my first SSH connection) and also store password (for the second SSH connection), which I can use later in the shell script? If there is no way to securely store passwords that is still fine, i can change the second SSH to use SSH keys, but not the other way around (I have no access to the Proxy).

Thanks,
Greg

**Comment by madhurig** [comment]
@gregkalapos: 

To secure passwords, you can define them as a secret variable on the definition in the "Variables" tab. If you use the "lock" icon next to the variable, it will be treated as a secret i.e. stored encrypted on the server and build machine, not logged in the output logs etc. 

Current options for private key:
1. Store the value in a "secret" variable on the definition
2. Encrypt and store it in source, we have tasks that you can use to decrypt at build time using the "Decrypt File" task. We have an example of using this task for [iOS signing certificates](https://www.visualstudio.com/en-us/docs/build/apps/mobile/secure-certs). I agree this may still not be acceptable. 

In the next VSTS update, we will release a preview feature that will allow you to store files like private keys securely on the server. One of the scenarios we are trying to address with that is SSH key installs and setup on the build agent. But it is not available right now.

**Comment by gregkalapos** [comment]
@madhurig Thanks for the feedback! 

The "Decrypt File" task works fine, that is great! The last problem I have is that in order to log in via SSH jump I have to enter 1) the passphrase to the SSH key and 2) the user password to the machine I deploy to (And of course in this case I have to script this). 

Now I found this: 
http://stackoverflow.com/questions/16928004/how-to-enter-ssh-password-using-bash 
but this expect script does not seem to be supported. 

Then I found this:
http://stackoverflow.com/questions/12202587/automatically-enter-ssh-password-with-script
But this needs additional installation on the agent. 

What is the way to go in such cases? I know this is rather related to scripting and not to VSTS, but I guess this is a general problem many people have. Is there a documentation on this? Or do you have some hints for me?

Thanks

**Comment by madhurig** [comment]
@gregkalapos: Interactive password input is problematic, that is one of the reasons we chose SSH2 native node module instead of ssh command line tools. 
I haven't used an expect script before, so I can't explain why that does not work. Are you using the Hosted build pool or your own machine? If the script requires interactive login, that might be theissue.

The current solution for users who cannot use our built in SSH tasks is to manually install the SSH keys on their private build servers, so their build scripts can run without requiring passphrase/password input prompts. We have plans to improve this and we have started work in the area but it is not at a point where you can use it as is. We are now shipping a "preview" task that will allow installing SSH keys stored securely on the server (no need to put keys in source code and encrypt) during the build: https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/InstallSSHKey. This task sets up the SSH agent, installs keys and cleans up after the build. But it does not support SSH keys with passphrase yet. 

Thanks,
Madhuri

**Comment by gregkalapos** [comment]
> Are you using the Hosted build pool or your own machine?

I use the hosted build pool.

Ok, so the only option I see as a workaround for my problem is this:

- I will remove the passphrase from the SHH key, which I use to log in to the Proxy 
- Currently I log in from the Proxy to the real deployment target with username/password. I will try to do this also with an SSH key (also without a passphrase). I'm not sure I can pass an SHH key to this connection from the build agent (I cannot store it on the Proxy-Machine, so both SSH keys will be on the build agent).

If I manage to do the steps above then I will be able to automatically deploy with a "Shell Script" + a "Decrypt File" task.

Some comments:

- The "Install SSH Key" task is nice, but actually solves a problem, which we already could workaround with the "Decrypt File" task (I was able to implement it very quickly, and it was not that bad). Of course if that integrates with the "Copy Files Over SSH" task and the "SSH" task then it is a real help.
- I think it would be better to extend the existing SSH tasks to be something like a real SSH client (for example like Putty). - Also with ProxyCommand support.

I think we discussed all options, so from my side: we can close this issue (unless you need it for future improvements). 

The bottom line is: it would be very good to have a real SSH client (real = completely implement the SSH protocol + a UI, where users can add parameters - like Putty) in VSTS.

Thanks for your help!

**Comment by gregkalapos** [comment]
Btw I created a screencast which shows my current setup and explains how I solved this problem: http://kalapos.net/Blog/ShowPost/cicdpipelinescreencast

**Comment by leantk** [comment]
This issue is stale because it has been open for a year with no activity. Remove the stale label or comment on the issue otherwise this will be closed in 5 days
